application/src/main/java/org/opentripplanner/street/search/state/EdgeTraverser.java

```diff
@@ -14,24 +17,49 @@
  */
 public class EdgeTraverser {
 
-  public static Optional<State> traverseEdges(final State s, final Collection<Edge> edges) {
-    var state = s;
+  public static Optional<State> traverseEdges(
+    final Collection<State> initialStates,
+    final Collection<Edge> edges
+  ) {
+    return traverseEdges(initialStates.toArray(new State[0]), edges);
+  }
+
+  public static Optional<State> traverseEdges(
+    final State[] initialStates,
+    final Collection<Edge> edges
+  ) {
+    if (edges.isEmpty()) {
+      return Optional.of(initialStates[0]);
+    }
+
+    // The shortest path tree is used to prune dominated parallel states. For example,
+    // CAR_PICKUP can return both a CAR/WALK state after each traversal of which only
+    // the optimal states need to be continued.
+    var dominanceFunction = new DominanceFunctions.MinimumWeight();
+    var spt = new ShortestPathTree<>(dominanceFunction);
+    for (State initialState : initialStates) {
+      spt.add(initialState);
+    }
+
+    Vertex lastVertex = null;
+    var isArriveBy = initialStates[0].getRequest().arriveBy();
     for (Edge e : edges) {
-      var afterTraversal = e.traverse(state);
-      if (afterTraversal.length > 1) {
-        throw new IllegalStateException(
-          "Expected only a single state returned from edge %s but received %s".formatted(
-              e,
-              afterTraversal.length
-            )
-        );
-      }
-      if (State.isEmpty(afterTraversal)) {
+      var vertex = isArriveBy ? e.getToVertex() : e.getFromVertex();
```

Comment:
Is the reason we need to care about arriveBy is that we now use the SPT?

Reply:
Yes, it is due to the SPT which stores states by the `vertex`. Previously we could ignore the from/to vertex handling since the single previous state was used.